libvast/include/vast/segment.hpp

```diff
@@ -31,6 +34,38 @@ class segment {
   friend segment_builder;
 
 public:
+  class iterator
+    : public detail::iterator_facade<
+        iterator, table_slice, std::random_access_iterator_tag, table_slice> {
+    friend detail::iterator_access;
+    using flat_slice_iterator = flatbuffers::Vector<
+      flatbuffers::Offset<fbs::FlatTableSlice>>::const_iterator;
+    using interval_iterator
+      = flatbuffers::Vector<const fbs::uinterval*>::const_iterator;
+
+  public:
+    iterator(flat_slice_iterator nested, interval_iterator intervals,
+             chunk_ptr chunk);
+
+    [[nodiscard]] table_slice dereference() const;
+
+    void increment();
+
+    void decrement();
+
+    void advance(size_t n);
+
+    [[nodiscard]] bool equals(iterator other) const;
+
+    [[nodiscard]] segment::iterator::difference_type
+    distance_to(iterator other) const;
+
+  private:
+    flat_slice_iterator nested_;
+    interval_iterator intervals_;
+    chunk_ptr chunk_;
```

Comment:
Why do we need the chunk? If we do let's at least add a comment that this is intentional, by default an iterator doesn't keep it's container alive.

Reply:
We could also keep a pointer to the segment itself and get the chunk from it when we construct the table slice.